crates/pathfinder/src/consensus/inner/p2p_task.rs

```diff
@@ -175,60 +198,84 @@ pub fn spawn(
-                            // sure occurred at least once during tests on my machine.
-                            tracing::warn!(
-                                "Engine requested gossiping a proposal for {height_and_round} via \
-                                 ConsensusEvent::Gossip but we did not create it due to missing \
-                                 respective ConsensusEvent::RequestProposal. my_proposals_cache: \
-                                 {my_proposals_cache:#?}, incoming_proposals_cache: \
-                                 {incoming_proposals_cache:#?}",
-                            );
-
-                            // The engine chose us for this round as proposer and requested that
-                            // we gossip a proposal from a
-                            // previous round.
-                            let mut prev_rounds_proposals = incoming_proposals_cache
-                                .remove(&proposal.height)
-                                .expect("Proposal was inserted into the cache");
-                            // For now we just choose the proposal from the previous round, and
-                            // the rest are kept for debugging
-                            // purposes.
-                            let (round, mut proposal_parts) = prev_rounds_proposals
-                                .pop_last()
-                                .expect("At least one proposal from a previous round");
-                            assert_eq!(
-                                round.as_u32().expect("Round not to be None") + 1,
-                                proposal.round.as_u32().expect("Round not to be None")
-                            );
-                            let ProposalInit {
-                                round, proposer, ..
-                            } = proposal_parts
-                                .first_mut()
-                                .and_then(ProposalPart::as_init_mut)
-                                .expect("First part to be Init");
-                            // Since the proposal comes from some previous round we need to
-                            // correct the round number and
-                            // proposer address.
-                            assert_ne!(
-                                *round,
-                                proposal.round.as_u32().expect("Round not to be None")
-                            );
-                            assert_ne!(*proposer, Address(proposal.proposer.0));
-                            *round = proposal.round.as_u32().expect("Round not to be None");
-                            *proposer = Address(proposal.proposer.0);
-                            proposal_parts
-                        };
+                        let consensus_storage2 = consensus_storage.clone();
+                        let consensus_storage3 = consensus_storage.clone();
```

Comment:
The amount of `storage.clone()` calls is fishy here: smells like bad design.

As far as I can see functions like `ValidatorFinalizeStage::validate()` and `commit_finalized_block()` (pretty much all functions accessing storage) take a `Storage` argument. That is effectively preventing transaction updates. That is, right now updating the Merkle tries is committed even if then inserting the block headers fail.

Instead of that, these functions fetching / updating storage state should take a transaction reference instead.

Reply:
Hmm, AFAIK some of them _did_ take a transaction reference, until recently... @CHr15F0x , do you remember?